src/Menu/PickerMenuBuilder.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Menu;
+
+use Contao\BackendUser;
+use Contao\CoreBundle\Event\BuildPickerMenuEvent;
+use Contao\CoreBundle\Event\ContaoCoreEvents;
+use Knp\Menu\FactoryInterface;
+use Knp\Menu\ItemInterface;
+use Knp\Menu\Renderer\RendererInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Creates the picker menu.
+ *
+ * @author Leo Feyer <https://github.com/leofeyer>
+ */
+class PickerMenuBuilder
+{
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $dispatcher;
+
+    /**
+     * @var FactoryInterface
+     */
+    private $factory;
+
+    /**
+     * @var RendererInterface
+     */
+    private $renderer;
+
+    /**
+     * @var string
+     */
+    private $pagePickerLabel = 'Page picker';
+
+    /**
+     * @var string
+     */
+    private $filePickerLabel = 'File picker';
+
+    /**
+     * Constructor.
+     *
+     * @param RequestStack             $requestStack
+     * @param RouterInterface          $router
+     * @param EventDispatcherInterface $dispatcher
+     * @param FactoryInterface         $factory
+     * @param RendererInterface        $renderer
+     */
+    public function __construct(RequestStack $requestStack, RouterInterface $router, EventDispatcherInterface $dispatcher, FactoryInterface $factory, RendererInterface $renderer)
+    {
+        $this->requestStack = $requestStack;
+        $this->router = $router;
+        $this->dispatcher = $dispatcher;
+        $this->factory = $factory;
+        $this->renderer = $renderer;
+    }
+
+    /**
+     * Sets the page picker label.
+     *
+     * @param string $pagePickerLabel
+     */
+    public function setPagePickerLabel($pagePickerLabel)
```

Comment:
I fully agree that the service must not be stateful.

However, I think it is ok if the service adds the two core navigation items. Do you really think we need a separate event listener for that?